drivers/sdhc/sdhc_stm32.c

```diff
@@ -0,0 +1,659 @@
+			return -EIO;
+		}
+
+		LOG_INF("Successfully switched to 1.8V signaling");
+	} else {
+		LOG_DBG("Card does not support 1.8V signaling");
+		return -ENOTSUP;
+	}
+
+	return res;
+}
+
+static int sdhc_stm32_request(const struct device *dev, struct sdhc_command *cmd,
+			      struct sdhc_data *data)
+{
+	int res;
+	struct sdhc_stm32_data *dev_data = dev->data;
+
+	__ASSERT(cmd != NULL, "Command is NULL.");
+
+	if (!WAIT_FOR(HAL_SDIO_GetState(&dev_data->hsd) == HAL_SDIO_STATE_READY, SD_TIMEOUT,
+		      k_usleep(1))) {
+		LOG_ERR("SD Card is busy now");
+		return HAL_ERROR;
+	}
+
+	(void)pm_device_runtime_get(dev);
+	/* Prevent the clocks to be stopped during the request */
+	pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+	k_mutex_lock(&dev_data->bus_mutex, K_FOREVER);
+
+	dev_data->cmd_index = cmd->opcode;
+	switch (cmd->opcode) {
+	case SD_GO_IDLE_STATE:
+		res = sdhc_stm32_go_idle_state(dev);
+		break;
+
+	case SD_SELECT_CARD:
+		res = SDMMC_CmdSelDesel(dev_data->hsd.Instance, cmd->arg);
+		break;
+
+	case SD_SEND_RELATIVE_ADDR:
+		res = SDMMC_CmdSetRelAdd(dev_data->hsd.Instance, (uint16_t *)&cmd->response);
+		break;
+
+	case SDIO_SEND_OP_COND:
+		res = SDMMC_CmdSendOperationcondition(dev_data->hsd.Instance, cmd->arg,
+						      (uint32_t *)&cmd->response);
+		break;
+
+	case SDIO_RW_DIRECT:
+		res = sdhc_stm32_rw_direct(dev_data, cmd);
+		break;
+
+	case SDIO_RW_EXTENDED:
+		res = sdhc_stm32_rw_extended(dev_data, cmd, data);
+		break;
+
+	case SD_VOL_SWITCH:
+		res = sdhc_stm32_switch_to_1_8v(dev);
+		break;
+
+	default:
+		res = -ENOTSUP;
+		LOG_DBG("Unsupported Command, opcode:%d\n.", cmd->opcode);
+		break;
+	}
+
+	if (res != HAL_OK) {
+		LOG_ERR("Command Failed, opcode:%d\n.", cmd->opcode);
+		sdhc_stm32_log_err_type(&dev_data->hsd);
+	}
+
+	k_mutex_unlock(&dev_data->bus_mutex);
+	(void)pm_device_runtime_put(dev);
+	pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+
+	return res;
+}
+
+static int sdhc_stm32_set_io(const struct device *dev, struct sdhc_io *ios)
+{
+	struct sdhc_stm32_data *data = dev->data;
+	struct sdhc_io *host_io = &data->host_io;
+	int res = 0;
+
+	(void)pm_device_runtime_get(dev);
+	/* Prevent the clocks to be stopped during the request */
+	pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+	k_mutex_lock(&data->bus_mutex, K_FOREVER);
+
+	if ((ios->clock != 0) && (host_io->clock != ios->clock)) {
+		if ((ios->clock > STM32_SDIO_F_MAX) || (ios->clock < STM32_SDIO_F_MIN)) {
+			LOG_ERR("Invalid Clock Frequency\n");
+			return -EINVAL;
+		}
+		res = HAL_SDIO_ConfigFrequency(&data->hsd, (uint32_t)ios->clock);
+		if (res != HAL_OK) {
+			LOG_ERR("Failed to set clock as %d\n", ios->clock);
+		} else {
+			host_io->clock = ios->clock;
+			LOG_INF("set SDHC clock as %d\n", ios->clock);
+		}
+		k_msleep(10);
+	}
+
+	if (ios->power_mode == SDHC_POWER_OFF) {
+		(void)SDMMC_PowerState_OFF(data->hsd.Instance);
+		k_msleep(10);
```

Comment:
This should be managed by `power-delay-ms`, as described above. In fact, we should likely be toggling the same power GPIO here